File: Practice/525.contiguous-array.java
public class Solution {
    public int findMaxLength(int[] nums) {
        Map<Integer, Integer> map = new HashMap<Integer, Integer>() {
            {
                put(0, 0);
            }
        };
        int maxLength = 0, runningSum = 0;
        for (int i = 0; i < nums.length; i++) {
            runningSum += nums[i];
            Integer prev = map.get(2 * runningSum - i - 1);
            if (prev != null)
                maxLength = Math.max(maxLength, i + 1 - prev);
            else
                map.put(2 * runningSum - i - 1, i + 1);
        }
        return maxLength;
    }
}

File: 525.contiguous-array.js
/*
 * @lc app=leetcode id=525 lang=javascript
 *
 * [525] Contiguous Array
 */

// @lc code=start
/**
 * @param {number[]} nums
 * @return {number}
 */
/* Solution 1: HashMap */
let findMaxLength = function (nums) {
  let map = {};
  map[0] = -1;
  let [maxlen, count] = [0, 0];

  for (let i = 0; i < nums.length; i++) {
    count = count + (nums[i] == 1 ? 1 : -1);
    if (map.hasOwnProperty(count)) {
      maxlen = Math.max(maxlen, i - map[count]);
    } else {
      map[count] = i;
    }
  }
  return maxlen;
};
// @lc code=end
